p2pserver/link/link.go

```diff
@@ -109,48 +109,62 @@ func (this *Link) Rx() {
 	for {
 		msg, payloadSize, err := types.ReadMessage(reader)
 		if err != nil {
-			log.Infof("[p2p]error read from %s :%s", this.GetAddr(), err.Error())
+			log.Infof("[p2p]error read from %s :%s", self.GetAddr(), err.Error())
 			break
 		}
 
-		t := time.Now()
-		this.UpdateRXTime(t)
+		self.time = time.Now()
 
-		if !this.needSendMsg(msg) {
-			log.Debugf("skip handle msgType:%s from:%d", msg.CmdType(), this.id)
+		if !self.needSendMsg(msg) {
+			log.Debugf("skip handle msgType:%s from:%d", msg.CmdType(), self.id)
 			continue
 		}
 
-		this.addReqRecord(msg)
-		this.recvChan <- &types.MsgPayload{
-			Id:          this.id,
-			Addr:        this.addr,
+		self.addReqRecord(msg)
+		self.recvChan <- &types.MsgPayload{
+			Id:          self.id,
+			Addr:        self.addr,
 			PayloadSize: payloadSize,
 			Payload:     msg,
 		}
 
 	}
 
-	this.CloseConn()
+	self.CloseConn()
 }
 
 //close connection
-func (this *Link) CloseConn() {
-	if this.conn != nil {
-		this.conn.Close()
-		this.conn = nil
+func (self *Link) CloseConn() {
+	self.sendBuf.TakeAndSeal()
+	if self.conn != nil {
+		_ = self.conn.Close()
+		self.conn = nil
 	}
 }
 
-func (this *Link) Send(msg types.Message) error {
+func (self *Link) Send(msg types.Message) error {
 	sink := comm.NewZeroCopySink(nil)
 	types.WriteMessage(sink, msg)
 
-	return this.SendRaw(sink.Bytes())
+	return self.SendRaw(sink.Bytes())
 }
 
-func (this *Link) SendRaw(rawPacket []byte) error {
-	conn := this.conn
+func (self *Link) SendAsync(msg types.Message) error {
+	sink := comm.NewZeroCopySink(nil)
+	types.WriteMessage(sink, msg)
+	return self.SendRawAsync(sink.Bytes())
+}
+
+func (self *Link) SendRawAsync(packet []byte) error {
+	if !self.sendBuf.Push(packet) {
+		return io.ErrClosedPipe
+	}
+
+	return nil
+}
+
+func (self *Link) SendRaw(rawPacket []byte) error {
+	conn := self.conn
```

Comment:
need process sendBuf firstly, to keep FIFO

Reply:
it is handled in link.go:237